flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java

```diff
@@ -1816,6 +1819,60 @@ public void testJobMasterOnlyTerminatesAfterTheSchedulerHasClosed() throws Excep
         jobMasterTerminationFuture.get();
     }
 
+    @Test
+    public void testJobMasterAcceptsExcessSlotsWhenJobIsRestarting() throws Exception {
+        configuration.set(RestartStrategyOptions.RESTART_STRATEGY, "fixed-delay");
+        configuration.set(
+                RestartStrategyOptions.RESTART_STRATEGY_FIXED_DELAY_DELAY, Duration.ofDays(1));
+        final JobMaster jobMaster =
+                new JobMasterBuilder(jobGraph, rpcService)
+                        .withConfiguration(configuration)
+                        .createJobMaster();
+
+        try {
+            jobMaster.start();
+
+            final JobMasterGateway jobMasterGateway =
+                    jobMaster.getSelfGateway(JobMasterGateway.class);
+
+            assertThat(
+                    jobMasterGateway.requestJobStatus(testingTimeout).get(), is(JobStatus.RUNNING));
+
+            final LocalUnresolvedTaskManagerLocation unresolvedTaskManagerLocation =
+                    new LocalUnresolvedTaskManagerLocation();
+            registerSlotsAtJobMaster(
+                    1,
+                    jobMasterGateway,
+                    jobGraph.getJobID(),
+                    new TestingTaskExecutorGatewayBuilder()
+                            .setAddress("firstTaskManager")
+                            .createTestingTaskExecutorGateway(),
+                    unresolvedTaskManagerLocation);
+
+            jobMasterGateway.disconnectTaskManager(
+                    unresolvedTaskManagerLocation.getResourceID(),
+                    new FlinkException("Test exception."));
+
+            assertThat(
+                    jobMasterGateway.requestJobStatus(testingTimeout).get(),
+                    is(JobStatus.RESTARTING));
```

Comment:
Need to retry this operation because there's no guarantee the jobmaster has processed the disconnect yet.

Reply:
I think in this specific case there is no race condition. But it relies on internal implementation details. I will add a retry.

Reply:
> in this specific case there is no race condition

Why is that? I thought that since the JM runs in an actual actor system without a DirectExecutor the processing of the disconnect can happen at some point in the future.

Reply:
The `jobMasterGateway` will serialize the messages that are sent over it. Hence, the JM will first receive the `disconnectTaskManager` message and then the `requestJobStatus`. Since the former message is implemented synchronously, the job will be in `RESTARTING` state when `requestJobStatus` is being processed.